### ISSUE TITLE
Concurrence issue in SelectableBiDirectionalQueue (package com.refinitiv.eta.valueadd.common) from the Java ETA API

### ISSUE THREAD

**Issue by mmd-financial-com** [Refinitiv/Real-Time-SDK#223]
Hi,

I think there is an issue with the way the _readLock and _writeLock are handled in the class SelectableBiDirectionalQueue (package com.refinitiv.eta.valueadd.common) of the Java ETA API.

Indeed, on the method read() (respectively write()) the _readLock (respectively the _writeLock) is locked. There is a finally clause that unlock the locks, however in case of an IOException, the shutdown method is called. This method will set to null the two respective locks. So if a thread was waiting on the lock, then it will never be able to continue its execution further. Moreover the other thread that sets the lock to null will run into a NullPointerException while executing the finally clause.

Best regards,
Marjorie.

**Comment by ayurov2** [comment]
Hi @mmd-financial-com, I appreciate you reporting this to us. I have opened an internal investigation JIRA to cover what you have found here.

**Comment by mmd-financial-com** [comment]
Hi,
In our code we are calling the shutdown function of the Reactor class and I think that about 1 time over 10 our application does not stop because of this issue. One thread is stuck in that function on the "while (!_esWorker.awaitTermination(SHUTDOWN_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS))" while one worker thread is stuck either on the write or the read of the SelectableBiDirectionalQueue.

**Comment by mmd-financial-com** [comment]
An easy way to reproduce the problem is to take the code https://github.com/Refinitiv/Real-Time-SDK/blob/master/Java/Eta/Applications/Examples/src/main/java/com/refinitiv/eta/examples/consumer/Consumer.java and to modify it slightly:  in the main don't call "consumer.uninitialize();". However at the beginning of the main, create a thread that has a reference on the consumer. On its run, this thread, will sleep an arbitrary number of seconds (between 1s and 45s) and then will call the "consumer.uninitialize();". 

You can create a script that will launch the consumer multiple times and you will see that from time to time the consumer does not stop.

**Comment by addicticks-dev** [comment]
Yes, @mmd-financial-com , I see a number of problems with [SelectableBiDirectionalQueue](https://github.com/Refinitiv/Real-Time-SDK/blob/master/Java/Eta/ValueAdd/src/main/java/com/refinitiv/eta/valueadd/common/SelectableBiDirectionalQueue.java).

-  `shutdown()` is being called _while holding the lock_. I don't think that is a good idea. The lock never gets cleared that way.
-  Inner class `NotifiedState` is said to be a way to "share a boolean between threads". But if that is the purpose of the class, then I think the boolean it holds should be declared `volatile`.
- In the `shutdown()` method I don't see the need to nullify the locks. Just let the GC do that. Nullifying just creates problems with read or write operations which may be in progress at the time. 
- The remote end is lazily created in the `remote()` method. However, this is also not thread safe. It needs to use the _double-checked locking_ method or if speed is not a concern: simply make the method `synchronized`.

**Comment by lbruun** [comment]
https://github.com/Refinitiv/Real-Time-SDK/commit/f0ec700373466617f4a686132df9a9279aafa166

**Comment by vlevendel** [comment]
Yes - addressed recently.  Please reopen or open a new issue if there are additional concerns.

**Comment by mmd-financial-com** [comment]
Thank you :)
